src/polyswarm_api/api.py

```diff
@@ -6,433 +7,427 @@
 
-    def rescan_submit(self, *hashes, **kwargs):
+    def search_by_feature(self, feature, *artifacts):
         """
-        Submit rescans to polyswarm and return UUIDs
+        Search artifacts by feature
 
-        :param artifact_type: What type to use when rescanning artifact
-        :param hashes: Hashable objects (Artifact, LocalArtifact, or Hash) or hex-encoded SHA256/SHA1/MD5
-        :return: SubmitResult generator
+        :param feature: Feature to use
+        :param artifacts: List of local.LocalArtifact objects
+        :return: Generator of ArtifactInstance resources
         """
-        hashes = [to_hash(h) for h in hashes]
-
-        futures = [(h, self.endpoint.rescan(h, **kwargs)) for h in hashes]
+        raise NotImplementedError()
 
-        for h, f in futures:
-            # artifact_type is not currently supported in rescan
-            yield result.SubmitResult(h, f.result(), self)
-
-    def scan(self, *artifacts):
+    def search_by_metadata(self, query):
         """
-        Submit artifacts to polyswarm and wait for scan results
+        Search artifacts by metadata
 
-        :param artifacts: List of LocalArtifacts or paths to local files
-        :return: ScanResult generator
+        :param query: A query string
+        :return: Generator of ArtifactInstance resources
         """
-        for submission in self.submit(*artifacts):
-            s = submission.wait_for_scan()
-            s.artifact = submission.artifact
-            yield s
+        return self.generator.search_metadata(query).execute().consume_results()
 
-    def rescan(self, *hashes, **kwargs):
+    def submit(self, artifact, artifact_type=resources.ArtifactType.FILE):
         """
-        Rescan artifacts via polyswarm
+        Submit artifacts to polyswarm and return UUIDs
 
-        :param hashes: Hashable objects (Artifact, LocalArtifact, or Hash) or hex-encoded SHA256/SHA1/MD5
-        :param kwargs: Keyword arguments for the scan (none currently supported)
-        :return: ScanResult generator
+        :param artifact: List of local.LocalArtifacts or paths to local files
+        :param artifact_type: The ArtifactType or strings containing "file" or "url"
+        :return: Generator of Submission resources
         """
-        for submission in self.rescan_submit(*hashes, **kwargs):
-            if submission.status_code == 200:
-                s = submission.wait_for_scan()
-                s.artifact = submission.artifact
-                yield s
+        if isinstance(artifact, string_types):
+            artifact_type = resources.ArtifactType.parse(artifact_type)
+            if artifact_type == resources.ArtifactType.FILE:
+                path = artifact
+                artifact_name = os.path.basename(artifact)
+                content=None
             else:
-                yield submission
+                path = None
+                artifact_name = artifact
```

Comment:
I'm confused by the logic here. I'm currently in a situation where I already have file content loaded from disk, and I want to submit it without writing it to a file. This is a fairly common situation; in some cases, users don't want to write something that is potentially malicious to disk.

The previous logic I believe accepted either a path or a `LocalArtifact`. If I try to manually create and pass a `LocalArtifact` here though, it will result in *another* one being created, with the content and the name both being set to the object, not appropriate values. If I pass the raw data in, a `LocalArtifact` is created, but the name is set to the artifact data, which is also definitely not what we want.

I think going back to the behavior where we can call this function with a path or an existing `LocalArtifact` object is the appropriate way to do this.

Reply:
The current logic assumes that you can provide a `LocalArtifact`. If you provide a `str`, it will consider it a path try to load from the file, if you provide a `bytes`, it will consider it the contents of the artifact and create the `LocalArtifact` with it, and if you provide a `LocalArtifact`, it will use it straight away.

It would be better, however, to accept a file-like. That way, you can build a `BytesIO` object or use a regular file if you wish. That is the default python abstraction for a file.I would prefer that the user provide either an path or file-like and then we would internally build the instance.

It also avoids making the user having to learn two signatures, one to build the `LocalArtifact` and another to submit it, and it makes things more straight-forward. Python people would expect to be able to provide native python objects. I would still accept a `LocalArtifact` for the specific case when the user gets that as a response from another api method and want to chain stuff, such as
```python
api.submit(api.download(api.lookup('some_hash')))
```
That, in my opinion, is a valid use case that justifies this feature. 

I was dealing with the case when the user provides the contents as an array of `bytes` directly, but that is not the best way really. I will add this to the changes so that we can provide a file-like and create the submission from it, it is much more clean. :+1: